src/java.base/share/classes/java/io/Reader.java

```diff
@@ -140,6 +140,208 @@ public void close() {
+                ensureOpen();
+                if (next >= length)
+                    return 0;
+                // Bound skip by beginning and end of the source
+                long r = Math.min(length - next, n);
+                r = Math.max(-next, r);
+                next += (int)r;
+                return r;
+            }
+
+            /**
+             * Tells whether this stream is ready to be read.
+             *
+             * @return True if the next read() is guaranteed not to block for input
+             *
+             * @throws     IOException  If the stream is closed
+             */
+            @Override
+            public boolean ready() throws IOException {
+                ensureOpen();
+                return true;
+            }
+
+            /**
+             * Tells whether this stream supports the mark() operation, which it does.
+             */
+            @Override
+            public boolean markSupported() {
+                return true;
+            }
+
+            /**
+             * Marks the present position in the stream.  Subsequent calls to reset()
+             * will reposition the stream to this point.
+             *
+             * @param  readAheadLimit  Limit on the number of characters that may be
+             *                         read while still preserving the mark.  Because
+             *                         the stream's input comes from a string, there
+             *                         is no actual limit, so this argument must not
+             *                         be negative, but is otherwise ignored.
+             *
+             * @throws     IllegalArgumentException  If {@code readAheadLimit < 0}
+             * @throws     IOException  If an I/O error occurs
+             */
+            @Override
+            public void mark(int readAheadLimit) throws IOException {
+                if (readAheadLimit < 0){
+                    throw new IllegalArgumentException("Read-ahead limit < 0");
+                }
+                ensureOpen();
+                mark = next;
+            }
+
+            /**
+             * Resets the stream to the most recent mark, or to the beginning of the
+             * string if it has never been marked.
+             *
+             * @throws     IOException  If an I/O error occurs
+             */
+            @Override
+            public void reset() throws IOException {
+                ensureOpen();
+                next = mark;
+            }
+
+            /**
+             * Closes the stream and releases any system resources associated with
+             * it. Once the stream has been closed, further read(),
+             * ready(), mark(), or reset() invocations will throw an IOException.
+             * Closing a previously closed stream has no effect.
+             */
+            @Override
+            public void close() {
+                cs = null;
+            }
```

Comment:
@AlanBateman I need to confess that I did not understand what you had in mind when you wrote this on the mailing list:
> That doesn't excuse you completely from thinking about concurrent use as Readers have a close method so you'll need to think about how close is specified for when it is called while another thread is reading chars from a custom CS.

As this new implementation explicitly is for single-threaded use only, there is no such other thread that could call `close` concurrently.

Maybe I am missing something here, so I would kindly ask for an outline of a scenario where -despite the explicit single-thread note- a second thread *does* exist?

Reply:
A scenario would be that this reader is shared by multiple threads in an argument to `Runnable`, like create 100 runnables, 50 are trying to close while 50 are reading to char arrays. There will be scenarios where some reading thread is reading while some closing thread invokes `close` and completes before the reading thread completes.

I think the key takeaway here is that `close` does not block until other threads finish reading, does not affect already-begun reading operations, and the closed state might not be immediately observed by other threads (as we do not release-write and acquire-read the cs clearing state). However, I don't know which of these attributes should be explicitly specified in our API specification.

Reply:
As Alan suggested, I just now in this very minute have added a more clear statement about the fact that it is invalid to share this Reader instance among multiple threads without *external* synchronization. That will prevent the scenario you mention.